### ISSUE TITLE
Feature: Add redundancy for data stored in the repository

### ISSUE THREAD

**Issue by fd0** [restic/restic#256]
Restic does its best to reduce the amount of redundant data in a backup, e.g. by saving duplicate chunks of files only once. When the repository files are damaged at the storage location (e.g. by bit rot), this means that a large number of files/metadata/snapshots are affected because of this reduced redundancy.

This issue tracks implementing something that artificially adds some redundancy back. I'm thinking along the lines of a reed-solomon error correcting code, e.g. as implemented by the famous `par2` program. This computes for a set of files (e.g. a large RAR archive with 100 individual `.rXX` files) a few `.par2` files which can in the case of data corruption be used to repair the original files to a certain amount. An introduction to the reed-solomon family of codes can be found here: https://www.backblaze.com/blog/reed-solomon/

A Go implementation is available here: https://github.com/klauspost/reedsolomon

**Comment by cfcs** [comment]
This would be a highly interesting feature to have if it was possible to distribute the shards to different targets (folders: different disks; hosts: offsite locations), allowing Tahoe-LAFS like RAID modes of efficient redundancy in scenarios where you can assume that ALL of your backup locations won't blow up at the same time, yet don't feel like making full, unneeded 1:1 backups of everything.

**Comment by cowai** [comment]
@cfcs Yes I agree. Configurable folders where each part should be stored would be totally awesome. That way I could use 5-6 disks without raid, and have the redundancy I need.

@fd0
Having a parity feature for restic would be the killer feature for me. I use attic today, but knowing that it will take care of bit rot would make me switch fast ;)

**Comment by Intensity** [comment]
What about STAIR codes?  http://ansrlab.cse.cuhk.edu.hk/software/stair/

**Comment by fd0** [comment]
Looks like something we could do, although I think it's much more complex than reed-solomon codes. I like the simplicity of the latter :)

**Comment by anarcat** [comment]
i just found out about SeqBox, which is an interesting approach to this:

https://github.com/MarcoPon/SeqBox

... instead of storing multiple redundant blocks, it tags each block with a specific metadata header so that the content can easily be found even after major filesystem or storage medium damage. it has really interesting properties, even though it looks a little like "magic":

![some dude making magic with his fingers](https://camo.githubusercontent.com/04d4093644aaf57077818a1faad2db57b675b4a3/687474703a2f2f692e696d6775722e636f6d2f44515a444f30502e676966)

the borg project also has an issue tracking various potential solutions for this in https://github.com/borgbackup/borg/issues/225 - and there's an interesting discussion there about the pros and cons of redundancy...

**Comment by cfcs** [comment]
SeqBox doesn't add redundancy; it enables you to carve out the puzzle pieces provided you have a block-aligned file system that lost its inode table but didn't lose any of the actual data. It does not help you recover from damaged storage blocks.
It does this at a massive overhead of ~3% - that's 30GB/TB.
If aligned to extents or to logical blocks (like - 1024) instead of 512 byte blocks the overhead would be smaller, but ... filesystems like ZFS are there to help resolve issues like this.

Given the nature of restic (writing very large volumes of data) I'd intuitively argue that there could be a larger statistical risk of bad blocks being data blocks than the inode index tables.

I think SeqBox is an interesting idea, but perhaps not a perfect match for restic's use-case.

**Comment by anarcat** [comment]
> It does not help you recover from damaged storage blocks.

I disagree. I think it can help recover damaged datasets with parts removed. It would allow someone to better recover if pieces of the filesystem went missing. Not entirely recover of course, but recover parts of the repo. How restic handles a partially destroyed repository is another thing, of course... But I think the comparison with PhotoRec is pretty convincing.

I do agree that a 3% overhead is huge, however. But when compared to par2, it's not that bad. :) Also keep in mind that the reed-solomon erasure coding algorithm proposed above talks about having 20-40% overhead, at least in the go library benchmarks. That's much larger, although the gains might be greater.

The question is: once the filesystem starts getting damaged, can you still recover enough the restic archive blocks? SeqBox helps finding those blocks and reed-solomon helps recovering the entirety of the original archive. I'm not sure you could recover the whole archive with only one of the two, unless reed-solomon is implemented at the filesystem level...

But you are right that, statistically, the odds are larger that data gets damaged than index. It's just that index data then becomes really precious: if any is lost, the result is complete catastrophe.

**Comment by cfcs** [comment]
I appreciate that it can help find blocks when the underlying FS is damaged and restic data has not been damaged.

I disagree that it can help solve the problem described in this issue post which is: Recover the original data put into restic in the scenario where restic data has been damaged:
>> When the repository files are damaged at the storage location [..]

IMHO comparing to `par2` is like comparing apples and oranges, they solve different problems.

I think we fundamentally agree on all the technical points, but disagree about the exact problem we're trying to solve in this issue.

Your third paragraph makes a good point, which I think should be the takeaway from the recent posts about `SeqBox` in this thread: `SeqBox`-style block/extent tagging alone does NOT help us recover from bitrot of the actual restic data, but combined with real error-correction it can be useful to help lay the puzzle.

It is however highly FS-dependent (block size could be different, FS could implement compression [like `NTFS`, `ZFS`, ...). I would still recommend relying on the FS layer for block indexing, and the main benefit of implementing application-layer parity would IMO be that it can be distributed across several physical storage backends.

That being said I appreciate you raising the point, it's an interesting subject to include in the discussion of an eventual error-correcting design. :-)

**Comment by cfcs** [comment]
Dumping this link here (from Hacker News today) for future reference: https://github.com/Bulat-Ziganshin/FastECC

**Comment by anarcat** [comment]
after some reflexion, i have come to believe that implementing error correction necessarily means the admins need to make decisions about where to store redundancy blocks. partial errors on drives is something that is quite uncommon: these days, especially with SSDs or SMART-enabled devices, drives often see total unrecoverable failures instead of partial ones.

what are we trying to protect against here? we should more clearly define what kind of failures we want to recover from before trying to find solutions. this is highly medium-dependent: a tape will fail completely differently from a HDD or a SDD, and different future-proof recovery mechanism should be implemented in those scenarios. are we defending against medium loss, physical damage or cosmic rays? those are also all very different...

borg previously pushed this issue aside and argued it is best delegated to the filesystem layer. i'm beginning to see a little more why.

concretely, i think reedsolomon would be interesting in the case where we don't trust the filesystem, e.g. if we don't want to rely on S3's reliability and want to distribute among *multiple* backends (different S3 enpoinds, S3/SSH or multiple drives or mixing it all)...

**Comment by cd1188** [comment]
btw ECC was the only thing to use RAR as a backup-tool for years, because no other archive-tool used that... its a good feeling to know, if a archive has a bitrot i can easy fix it with 3% of recovery-data in the rar-file

but i think its also better to use a minimum of 2 locations of your backup.
how does restic handle an error in the repro? 
does restic then heal the faulty data from the local source?
and what if it is no longer available locally, because the error was in an older snapshot?
maybe it would be also good if you could use a second repository to 'HEAL' the first repository?
like a fallback-repo :-)

i have another idea/question.
what if you read at backup-time unrecognized faulty data (sata-error / hdd-error / ram-error)?
in that case it would be also good to make always a second backup in a second-run in another repository without that errors.
-> by the way that could be a downside if you make a backup to two repos at the same time.
-> reading data twice would be safer and hopefully reading from disk and not from cache-buffer ;-)

does restic recognizes on the next snapshot, that in the first repository, the source-data does not match? because the 'change date' from the file are the same but the data are not?
i think the command 'restic check --read-data' checks only the repository and not against the original-source?

**Comment by igpit** [comment]
+1 for adding some ECC.

ofc you should use filesystems with checksumming, snapshots and redundancy underneath. but you can still for whatever reason end up with a file where just one sectorcannot be read correctly or a few bytes turn out incorrect. and you might have to read from a DVD where someone copied files to...

integrating something like par2 doesn't seem too complicated and can save your life. it could be optional as it creates only additional files. no harm done when missing.

if implemented, it should be done so that par2-files are created on the fly while writing the regular pack files. this way there is no 2nd pass needed to read already written data.

**Comment by iluvcapra** [comment]
+1

> if implemented, it should be done so that par2-files are created on the fly while writing the regular pack files. this way there is no 2nd pass needed to read already written data.

I'm watching for this feature myself. Though any kind of ECC should be robust enough to handle a couple different cases, like:

- 0.1% of all of the bytes in all of the packs is corrupt. (1 par file with 10% redundancy for each pack would easily be able to fix this.)
- 0.1% of all of the packs have been entirely lost. (1 par file with 10% redundancy for each pack would not fix this.)

**Comment by dumblob** [comment]
Please see my recent comment https://github.com/restic/restic/issues/804#issuecomment-615140267 before trying to come up with yet another **flawed** FEC (forward error correction).
